### ISSUE TITLE
About Class Cropr

### ISSUE THREAD

**Issue by Afleve** [benbergner/cropr#2]
Thank you for your contribution. I have encountered the following problem:
My task is to classify birds, and I directly applied your Crop module to the last layer of Vit. But I reconstructed the original image based on the preserved patches and found that it would obscure the target area.
Another confusing point is that in classification tasks, q is equivalent to a vector (N=1). This vector gives higher scores to patches it considers meaningful, which is implemented using dot product (q @ k. T) in the code. Can I assume that patches with high scores from different samples selected by q also have high similarity, even if cross entropy loss is used for classification in downstream tasks.
Looking forward to your reply.

**Comment by Afleve** [comment]
![1736514274828](https://github.com/user-attachments/assets/dfb4533e-4675-459c-9b83-cc68540fe77a)

**Comment by Afleve** [comment]
![image](https://github.com/user-attachments/assets/2cb7e198-153d-4a1b-907a-19da9dc1eca0)

**Comment by benbergner** [comment]
Thanks a lot for trying Cropr on a fine-grained classification task.

> My task is to classify birds, and I directly applied your Crop module to the last layer of Vit.

I assume you want to use Cropr for interpretability purposes, i.e. to identify task-important image patches, since you only apply it to the last layer? If you want to accelerate your ViT, you would have to apply it earlier in the network.

> But I reconstructed the original image based on the preserved patches and found that it would obscure the target area.

There can be several reasons why the plot may not look right in general, but it is unusual to have so many of the task-important patches removed. Could you check if everything is correct in your plot generating function?

Some more general considerations:
- The CLS token, which is not to be visualized, needs to be treated accordingly.
- Since you are applying Cropr to the last layer, the auxiliary loss should be very similar to the main task loss after training.
- After applying Cropr, the tokens/patches are ordered according to their attention score and thus their importance. Since the CLS token score is set to infinity, it will always be positioned first. However, one must keep track of the original order of the patch tokens for plotting. This is not included in `cropr/cls` because the order of patches is not important for classification. However, you can check the code in `cropr/segm` where I keep track of the original patch order since it is important for segmentation.

Hope this helps with your debugging. Let me know if you need further help.

> Another confusing point is that in classification tasks, q is equivalent to a vector (N=1). This vector gives higher scores to patches it considers meaningful, which is implemented using dot product (q @ k. T) in the code. Can I assume that patches with high scores from different samples selected by q also have high similarity

In classification, q is indeed a vector. It is also true that after training, q will align well with tokens that are important for the task.

Your question is very interesting. Score logits (pre-softmax) could potentially also be useful for determining semantic similarity between patches from different images. However, this is something that needs to be explored further.

**Comment by Afleve** [comment]
Thank you for your reply. I think there is no problem with my image reconstruction code because I also used the attention value of the last layer of Vit itself to select patches. The two resulted in completely different outcomes, so I'm wondering if this Crop cannot be directly used to select patches after the last layer of Block.
![1736519429868](https://github.com/user-attachments/assets/17444934-eb7a-4502-8c26-ca08ca0e0543)

**Comment by benbergner** [comment]
Cropr is used across different layers in the paper, so it should work in your setting. Is the Cropr module sufficiently trained, i.e. is the auxiliary loss similarly low compared to the main task loss?

**Comment by Afleve** [comment]
> Cropr is used across different layers in the paper, so it should work in your setting. Is the Cropr module sufficiently trained, i.e. is the auxiliary loss similarly low compared to the main task loss?

yeah, I have already trained this auxiliary head. But this result does seem a bit unsatisfactory

**Comment by benbergner** [comment]
The more details you provide, the better I can help :-)

What exactly means unsatisfactory? What is the training loss of the auxiliary head in Cropr vs. main head? Could you share code to reproduce your results?

**Comment by Afleve** [comment]
I think I have found the reason, but I am quite confused.
My classification head consists of a normalization operation and a weight normalization linear classifier
The classification header in your code consists of a layer normalization and a linear classifier
I can get normal results by replacing your classification header in my code, but if I replace it with mine, the model will crash. I don't understand why

**Comment by Afleve** [comment]
I further conducted verification experiments and found that normalizing the features before my classifier would cause the entire model to be abnormal. I am a bit confused as to why this is the case
